### PR TITLE
Updates for ReaderScreen

### PR DESCRIPTION
**Pull request by revonateB0T**

### CODE REVIEW

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/GalleryPager.kt [hunk 1]
```diff
@@ -41,7 +46,18 @@ fun GalleryPager(
                 reverseLayout = type == RIGHT_TO_LEFT,
                 key = { it },
             ) { index ->
-                item(items[index])
+                val page = items[index]
+                val zoomableState = rememberZoomableState(zoomSpec = zoomSpec)
+                Box(
+                    modifier = Modifier.fillMaxSize().zoomable(
+                        state = zoomableState,
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/reader/GalleryPager.kt:53]: Content shifts if zoomed with two finger.

**Comment by FooIbar** [comment]
Also I got this while zooming with gestures.
```
java.lang.NullPointerException
	at me.saket.telephoto.zoomable.RealZoomableState.fling-BMRW4eQ$zoomable_release(RealZoomableState.kt:502)
	at me.saket.telephoto.zoomable.ZoomableNode$onTransformStopped$1$1.invokeSuspend(Zoomable.kt:138)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
	at androidx.compose.ui.platform.AndroidUiDispatcher.performTrampolineDispatch(AndroidUiDispatcher.android.kt:81)
	at androidx.compose.ui.platform.AndroidUiDispatcher.access$performTrampolineDispatch(AndroidUiDispatcher.android.kt:41)
	at androidx.compose.ui.platform.AndroidUiDispatcher$dispatchCallback$1.run(AndroidUiDispatcher.android.kt:57)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8294)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:977)
	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [androidx.compose.ui.platform.MotionDurationScaleImpl@f42ecf2, androidx.compose.runtime.BroadcastFrameClock@8dccf43, StandaloneCoroutine{Cancelling}@dfcd9c0, AndroidUiDispatcher@d24ff37]
```

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt [hunk 2]
```diff
@@ -82,7 +90,6 @@ fun ReaderScreen(info: BaseGalleryInfo, page: Int = -1) {
     Deferred({ pageLoader.awaitReady() }) {
         val lazyListState = rememberLazyListState()
         val pagerState = rememberPagerState { pageLoader.size }
-        val zoomableState = rememberZoomableState(zoomSpec = ZoomSpec(maxZoomFactor = 3f))
```
Comment (FooIbar) [inline app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: We may want it here for https://saket.github.io/telephoto/zoomable/#edge-detection

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt [hunk 3]
```diff
@@ -134,12 +160,8 @@ fun ReaderScreen(info: BaseGalleryInfo, page: Int = -1) {
                             Image(
```
[the remark this answers is not in the document]
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:162]: This should be fixed in PageLoader side

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt (hunk 2 above)
Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: Pager should only be zoomable when image is ready.

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/GalleryPager.kt (hunk 1 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/GalleryPager.kt:57]: Any idea how we fix this?

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/GalleryPager.kt:57]: Looks like it's on compose foundation side.

**Comment by revonateB0T** [comment]
Fuck, github又炸了 push的commit同步不过来了

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt (hunk 2 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: In WEBTOON, CONTINUOUS_VERTICAL mode, this is only one zoomState and globally zoomable
In DEFAULT, LEFT_TO_RIGHT, RIGHT_TO_LEFT, VERTICAL we should have multiple zoomState but currently, one
Hard to implement this

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: Pass page state to GalleryPager?

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: I mean, we shouldn't do this now,
As we currently use a global zoomState so it is keeped during page swipe,
User must be able to reset the scale state regardless of page state

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: That's another thing.
I was talking about the zoomable modifier should be applied on the image, progress indicator and error message shouldn't be zoomable.

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: Yeah, I agree that in DEFAULT, LEFT_TO_RIGHT, RIGHT_TO_LEFT, VERTICAL, they shouldn't be zoomable,
But in WEBTOON, CONTINUOUS_VERTICAL, they should have the scale per image

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: Let's implement it after https://github.com/FooIbar/EhViewer/issues/609

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: > Yeah, I agree that in DEFAULT, LEFT_TO_RIGHT, RIGHT_TO_LEFT, VERTICAL, they shouldn't be zoomable, But in WEBTOON, CONTINUOUS_VERTICAL, they should have the scale per image

Fair.

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: > > Yeah, I agree that in DEFAULT, LEFT_TO_RIGHT, RIGHT_TO_LEFT, VERTICAL, they shouldn't be zoomable, But in WEBTOON, CONTINUOUS_VERTICAL, they should have the scale per image
> 
> Fair.

Is https://github.com/revonateB0T/EhViewer/commit/eb7b8d5d29e7b7f25da9433fae0bdd8f8cf3fcb8 the one you mean?

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: > > > Yeah, I agree that in DEFAULT, LEFT_TO_RIGHT, RIGHT_TO_LEFT, VERTICAL, they shouldn't be zoomable, But in WEBTOON, CONTINUOUS_VERTICAL, they should have the scale per image
> > 
> > 
> > Fair.
> 
> Is [revonateB0T@eb7b8d5](https://github.com/revonateB0T/EhViewer/commit/eb7b8d5d29e7b7f25da9433fae0bdd8f8cf3fcb8) the one you mean?

That looks overcomplicated.
What about moving `item` to a separate Composable that takes something like `isWebtoon`?

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:162]: Done

File: app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt (hunk 2 above)
Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: > > > > Yeah, I agree that in DEFAULT, LEFT_TO_RIGHT, RIGHT_TO_LEFT, VERTICAL, they shouldn't be zoomable, But in WEBTOON, CONTINUOUS_VERTICAL, they should have the scale per image
> > > 
> > > 
> > > Fair.
> > 
> > 
> > Is [revonateB0T@eb7b8d5](https://github.com/revonateB0T/EhViewer/commit/eb7b8d5d29e7b7f25da9433fae0bdd8f8cf3fcb8) the one you mean?
> 
> That looks overcomplicated. What about moving `item` to a separate Composable that takes something like `isWebtoon`?

Then we have to make `zoomableState` controlled in different `Composable`

Comment (FooIbar) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: > > > > > Yeah, I agree that in DEFAULT, LEFT_TO_RIGHT, RIGHT_TO_LEFT, VERTICAL, they shouldn't be zoomable, But in WEBTOON, CONTINUOUS_VERTICAL, they should have the scale per image
> > > > 
> > > > 
> > > > Fair.
> > > 
> > > 
> > > Is [revonateB0T@eb7b8d5](https://github.com/revonateB0T/EhViewer/commit/eb7b8d5d29e7b7f25da9433fae0bdd8f8cf3fcb8) the one you mean?
> > 
> > 
> > That looks overcomplicated. What about moving `item` to a separate Composable that takes something like `isWebtoon`?
> 
> Then we have to make `zoomableState` controlled in different `Composable`

The item Composable can have a default zoomableState parameter.
And we pass the parent‘s zoomableState in webtoon scenario.

Comment (revonateB0T) [reply app/src/main/kotlin/com/hippo/ehviewer/ui/reader/ReaderScreen.kt:85]: fa05bd75c2dad787ceefd94a543546e863a90b3f
Then we have to pass `onMenuRegionClick`, `onSelectPage` to PagerItem
And conditionally wrap with a `Box` or make `PagerItem` a local `@Composable`
I don't think it is much cleaner

**Review by FooIbar — APPROVED** [review]

**Pull request merged**